Scrapers (#111)
hpe-storage/scod: docs/csi_driver/metrics.md

# HPE CSI Info Metrics Provider for Prometheus

The HPE CSI Driver for Kubernetes may be accompanied by a Prometheus metrics endpoint to provide metadata about the volumes provisioned by the CSI driver and supporting backends. It's conventionally deployed with [HPE Storage Array Exporter for Prometheus](https://hpe-storage.github.io/array-exporter) to provide a richer set of metrics from the backend storage systems.

[TOC]

## Metrics Provided

The exporter provides two metrics, "hpestoragecsi_volume_info" and "hpestoragecsi_backend_info".

### Volume Info

| Metric                    | Type    | Description                                                 | Value |
| :------------------------ | :------ | :---------------------------------------------------------- | :---- |
| hpestoragecsi_volume_info | Gauge   | Indicates a volume whose provisioner is the HPE CSI Driver. | 1     |

> 

Reviewer: Was this sentence omitted intentionally?  "This metric includes the following labels."

| Label         | Description                                                |
| :------------ | :--------------------------------------------------------- |

> | backend       | Backend hostname or IP address as defined in the `Secret`. |

Reviewer: The period can be removed, for consistency with the other label descriptions.

Author: I opted for period everywhere.

# HPE CSI Info Metrics Provider

The HPE CSI Driver for Kubernetes may be accompanied by a Prometheus metrics endpoint to provide metadata about the volumes provisioned by the CSI driver and supporting backends. It's conventionally deployed with [HPE Storage Array Exporter for Prometheus](https://hpe-storage.github.io/array-exporter) to provide a richer set of metrics from the backend storage systems.

[TOC]

## Exported Metrics

> The exporter provide two metrics, "hpestoragecsi_volume_info" and "hpestoragecsi_backend_info".

Reviewer 2: provides

| pvc_namespace | `PersistentVolumeClaim` `Namespace`                        |
| storage_class | `StorageClass` used to provision the `PersistentVolume`    |
| volume        | Backend volume name                                        |

### Backend Info

| Metric | Type | Labels | Description |
| :------------------------- | :----- | :------ | :------------------------------------------------------------------------ |
| hpestoragecsi_backend_info | Gauge  | backend | Indicates a storage system for which the HPE CSI driver is a provisioner. |

The labels have the following meanings.

| Label | Description |
| :------ | :-------------------------------------------------------- |
| backend | Backend hostname or IP address as defined in the `Secret` |

## Deployment

The exporter may be installed either via Helm or through YAML manifests with the object definitions. It's recommended to use Helm as it's more convenient to manage the configuration of the deployment.

!!! note
    It's recommended to add a "cluster" target label to the deployment. The label is used in the [provided Grafana dashboards](https://grafana.com/orgs/hpestorage/dashboards).

### Helm

> The Helm chart is available on Artifact Hub. All direction on how to manage and install the chart is available within the chart documentation.

Reviewer 2: ```suggestion
The Helm chart is available on Artifact Hub. Instructions on how to manage and install the chart is available within the chart documentation.
```

- [HPE CSI Info Metrics Provider for Prometheus Helm chart](https://artifacthub.io/packages/helm/hpe-storage/hpe-csi-info-metrics)

### Advanced Install

> Before beginning an advanced install, it needs to be determined how Prometheus is deployed on the Kubernetes cluster as it dictates how the scrape target is configured, either via a `Service` annotation or a `ServiceMonitor` CRD.

Reviewer 2: ```suggestion
Before beginning an advanced install, determine how Prometheus will be deployed on the Kubernetes cluster as it will dictate how the scrape target will be configured with either a `Service` annotation or a `ServiceMonitor` CRD.
```

> Start by downloading the manifest, it needs local changes before applying to the cluster.

Reviewer 2: ```suggestion
Start by downloading the manifest, it will need to be modified before applying to the cluster.
```

Author: ![image](https://user-images.githubusercontent.com/7026024/129981700-a8f38efb-e914-4ff0-bb17-24d762d1dbad.png)

Reviewer: Other options are to change the comma to a period or to change "it" to "which".

#### Version 1.0.0

Supports HPE CSI Driver for Kubernetes 2.0.0 and later.

```markdown
wget https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/csi-info-metrics/v1.0.0/hpe-csi-info-metrics.yaml
```

Optional, `ServiceMonitor` definition:

```markdown
wget https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/csi-info-metrics/v1.0.0/hpe-csi-info-metrics-service-monitor.yaml
```

#### Configuring Advanced Install

> The sections that needs be updated is the main container parameters and the optional annotation and label.

Reviewer 2: ```suggestion
The sections that need be updated are the main container parameters, the optional annotation and the label.
```

Reviewer: Label changes are also optional.  I suggest:

"Update the main container parameters and optionally add service labels and annotations."

In the "hpe-csi-info-metrics" `Deployment` at `.spec.template.spec.containers[0].args` in "hpe-csi-info-metrics.yaml":

```markdown
          args:
            - "--telemetry.addr=:9099"
            - "--telemetry.path=/metrics"
            # IMPORTANT: Uncomment this argument to confirm your
            # acceptance of the HPE End User License Agreement at
            # https://www.hpe.com/us/en/software/licensing.html
            #- "--accept-eula"
```

Remove the `#` in front of `--accept-eula` to accept the [HPE license restrictions](https://www.hpe.com/us/en/software/licensing.html).

In the "hpe-csi-info-metrics-service" `Service`:

```markdown
metadata:
  name: hpe-csi-info-metrics-service
  namespace: hpe-storage
  labels:
    app: hpe-csi-info-metrics
    # Optionally add labels, for example to be included in Prometheus
    # metrics via a targetLabels setting in a ServiceMonitor spec
    #cluster: my-cluster
  # Optionally add annotations, for example to configure it as a
  # scrape target when using the Prometheus Helm chart's default
  # configuration.
  #annotations:
  #  "prometheus.io/scrape": "true"
```

> - Apply and uncomment any custom labels. It's recommended to use a "cluster" label to use the [provided Grafana dashboards](https://grafana.com/orgs/hpestorage/dashboards).

Reviewer 2: ```suggestion
- Apply and uncomment any custom labels. It's recommended to use a "cluster" label with the [provided Grafana dashboards](https://grafana.com/orgs/hpestorage/dashboards).
```

- If Prometheus has been deployed without the Operator, uncomment the annotation.

Apply the manifest:

```markdown
kubectl apply -f hpe-csi-info-metrics.yaml
```

> Optionally, if using the Prometheus Operator, add any additional lables in "hpe-csi-info-metrics-service-monitor.yaml":

Reviewer 2: ```suggestion
Optionally, if using the Prometheus Operator, add any additional labels in "hpe-csi-info-metrics-service-monitor.yaml":
```
